typing/types.ml

```diff
@@ -135,36 +135,70 @@ and method_privacy =
   | Mprivate of field_kind
 
 (* Variance *)
+(* Variance forms a product lattice of the following partial orders:
+     0 <= may_pos <= pos
+     0 <= may_weak <= may_neg <= neg
+     0 <= inj
+   Additionally, the following implications are valid
+     pos => inj
+     neg => inj
+   Examples:
+     type 'a t        : may_pos + may_neg + may_weak
+     type 'a t = 'a   : pos
+     type 'a t = 'a -> unit : neg
+     type 'a t = ('a -> unit) -> unit : pos + may_weak
+     type 'a t = A of (('a -> unit) -> unit) : pos
+     type +'a p = ..  : may_pos + inj
+     type +!'a t      : may_pos + inj
+     type -!'a t      : may_neg + inj
+     type 'a t = A    : inj
+ *)
 
 module Variance = struct
   type t = int
   type f = May_pos | May_neg | May_weak | Inj | Pos | Neg | Inv
   let single = function
     | May_pos -> 1
-    | May_neg -> 2
+    | May_neg -> 2 + 4
     | May_weak -> 4
     | Inj -> 8
-    | Pos -> 16
-    | Neg -> 32
-    | Inv -> 64
+    | Pos -> 16 + 8 + 1
+    | Neg -> 32 + 8 + 4 + 2
+    | Inv -> 63
   let union v1 v2 = v1 lor v2
   let inter v1 v2 = v1 land v2
   let subset v1 v2 = (v1 land v2 = v1)
   let eq (v1 : t) v2 = (v1 = v2)
-  let set x b v =
-    if b then v lor single x else  v land (lnot (single x))
+  let set x v = union v (single x)
+  let set_if b x v = if b then set x v else v
   let mem x = subset (single x)
   let null = 0
   let unknown = 7
-  let full = 127
-  let covariant = single May_pos lor single Pos lor single Inj
-  let swap f1 f2 v =
-    let v' = set f1 (mem f2 v) v in set f2 (mem f1 v) v'
-  let conjugate v = swap May_pos May_neg (swap Pos Neg v)
+  let full = single Inv
+  let covariant = single Pos
+  let swap f1 f2 v v' =
+    set_if (mem f2 v) f1 (set_if (mem f1 v) f2 v')
+  let conjugate v =
+    let v' = inter v (union (single Inj) (single May_weak)) in
+    swap Pos Neg v (swap May_pos May_neg v v')
+  let compose v1 v2 =
+    if mem Inv v1 && mem Inj v2 then full else
+    let mp =
+      mem May_pos v1 && mem May_pos v2 || mem May_neg v1 && mem May_neg v2
+    and mn =
+      mem May_pos v1 && mem May_neg v2 || mem May_pos v1 && mem May_neg v2
+    and mw = mem May_weak v1 && v2 <> null || v1 <> null && mem May_weak v2
+    and inj = mem Inj v1 && mem Inj v2
+    and pos = mem Pos v1 && mem Pos v2 || mem Neg v1 && mem Neg v2
+    and neg = mem Pos v1 && mem Neg v2 || mem Neg v1 && mem Pos v2 in
+    List.fold_left (fun v (b,f) -> set_if b f v) null
+      [mp, May_pos; mn, May_neg; mw, May_weak; inj, Inj; pos, Pos; neg, Neg]
```

Comment:
This is actually rather readable, thanks!

I still wonder, is the treatment of `Inv` correct? I'm not sure what `Inv` is, but I would expect for example that composing `Pos` and `Inv` could be `Inv`? (Is this an emergent property of the computation below?)

Reply:
This is: `Inv` decompose into `Pos` and `Neg`, so that you would get again `Pos` and `Neg`.

Reply:
This gives me the impression again that `Inv` needs not be an element of `f`.

Reply:
It is there just as a short-hand of the union of `Pos` and `Neg`, as you can see in the fact it is not used in the result. We could also write `v1 = full`, but it makes more assumptions about the representation.

Reply:
My impression is that by removing `Inv`, we could express the same thing with less special cases and thus simpler code. Is there any downside to *not* having `Inv` in the set of atoms?